docs/advanced-topics/stale-index.md

```diff
@@ -0,0 +1,176 @@
+# Permanent stale index inconsistency
+
+In some situations due to crashes of storage database, index database, or JanusGraph instances 
+permanent stale index may appear.  
+One of such cases could be a vertex removal from the graph but due to a crash during index persistence 
+there is a chance that index won't be updated ever. 
+In case we try to remove a vertex which is already gone from the storage database but which is still indexed then 
+the `IllegalStateException` will be thrown with the message `Vertex with id %vertexId% was removed.`. 
+For example, using `g.V().has("name", "HelloWorld").drop().iterate(); g.tx().commit();` we may receive an exception 
+noting that some vertex has already been removed nevertheless it's record is still in the index. 
+The exception will be thrown anytime we try to remove such vertex from the graph.  
+This problem is known and should be temporal limitation until this issue is fixed in JanusGraph.  
+As for now there are several workarounds described below to fix permanent stale indices.  
+
+## StaleIndexRecordUtil
+
+`StaleIndexRecordUtil.class` is available in `janusgraph-core` module and is meant to be used as a helper class 
+to fix permanent stale index entries. 
+
+`StaleIndexRecordUtil.forceRemoveElementFromGraphIndex` can be used to force remove an entry of any element from 
+a graph index. Right now the limitations of this method is that it can be used with Composite indices only.  
+An example of using this method is below:
+```java
+StandardJanusGraph graph = (StandardJanusGraph) JanusGraphFactory.open(configuration);
+StandardJanusGraphTx tx = (StandardJanusGraphTx) graph.newTransaction();
+JanusGraphManagement mgmt = graph.openManagement();
+
+// Let's say we want to remove non-existent vertex from a stale index. 
+// We will assume the next constraints: 
+// vertex id is 12345;
+// Composite index name is: nameIndex
+// There is a single indexed property: name
+// Value of the vertex property is: HelloWorld
+
+long vertexId = 12345L;
+String compositeIndexName = "nameIndex";
+String propertyKeyName = "name";
+String value = "HelloWorld";
+
+PropertyKey propertyKey = mgmt.getPropertyKey(propertyKeyName);
+long propertyKeyId = propertyKey.longId();
+IndexRecordEntry namePropertyIndexRecord = new IndexRecordEntry(propertyKeyId, value, propertyKey);
+IndexRecordEntry[] fullIndexRecord = new IndexRecordEntry[]{namePropertyIndexRecord};
+
+JanusGraphElement elementToBeRemoved = new CacheVertex(tx, vertexId, ElementLifeCycle.New);
+
+// After the below method is executed index entry of the vertex 12345 should be removed from the index which 
+// effectively fixes permanent stale index inconsistency
+StaleIndexRecordUtil.forceRemoveElementFromGraphIndex(
+    elementToBeRemoved,
+    fullIndexRecord,
+    graph,
+    compositeIndexName
+);
+tx.commit();
+mgmt.rollback();
+```
+
+## Manually mutate index record
```

Comment:
My view is biased but I think this method is too complicated for normal users. It might benefit some people but other people can get confused and "scared". IMO It might be a better idea to post it on the mailing list or personal blog. Again, this is my personal opinion and does not stand for a rejection.

Reply:
I actually agree with you here. What do you think if we move this part of documentation into Javadoc of `StaleIndexRecordUtil` class?

Reply:
I moved all in-depth solutions into JavaDoc of `StaleIndexRecordUtil`. I left only an example usage of a simplified method `StaleIndexRecordUtil.forceRemoveElementFromGraphIndex(vertexId, propertyKeyName, value, graph, compositeIndexName)` in JanusGraph documentation. Let me know what you think about it